tools/istio-iptables/pkg/cmd/run.go

```diff
@@ -236,17 +234,31 @@ func (iptConfigurator *IptablesConfigurator) handleInboundIpv6Rules(ipv6RangesEx
 		// ::6 is bind connect from inbound passthrough cluster
 		iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "-s", "::6/128", "-j", constants.RETURN)
 
-		// Redirect app calls to back itself via Envoy when using the service VIP or endpoint
-		// address, e.g. appN => Envoy (client) => Envoy (server) => appN.
-		iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "!", "-d", "::1/128", "-j", constants.ISTIOINREDIRECT)
-
 		for _, uid := range split(iptConfigurator.cfg.ProxyUID) {
+			// Redirect app calls back to itself via Envoy when using the service VIP
+			// e.g. appN => Envoy (client) => Envoy (server) => appN.
+			// nolint: lll
+			iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "!", "-d", "::1/128", "-m", "owner", "--uid-owner", uid, "-j", constants.ISTIOINREDIRECT)
+
+			// Do not redirect app calls to back itself via Envoy when using the endpoint address
+			// e.g. appN => appN by lo
+			iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "-m", "owner", "!", "--uid-owner", uid, "-j", constants.RETURN)
+
 			// Avoid infinite loops. Don't redirect Envoy traffic directly back to
 			// Envoy for non-loopback traffic.
 			iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-m", "owner", "--uid-owner", uid, "-j", constants.RETURN)
 		}
 
 		for _, gid := range split(iptConfigurator.cfg.ProxyGID) {
+			// Redirect app calls back to itself via Envoy when using the service VIP
+			// e.g. appN => Envoy (client) => Envoy (server) => appN.
+			// nolint: lll
+			iptConfigurator.iptables.AppendRuleV6(constants.ISTIOOUTPUT, constants.NAT, "-o", "lo", "!", "-d", "::1/128", "-m", "owner", "--gid-owner", gid, "-j", constants.ISTIOINREDIRECT)
```

Comment:
Does it work as your comment?
when will the below condition hit at the same time?
1) `-o lo`
2) ! -d ::1/128
you mean you have a ip binded on lo but not ::1 ? I know ::6 is, but ::6 is not the service VIP you are mentioning

Reply:
Without thought that much, just covered case before. https://github.com/istio/istio/pull/19308/files#diff-7737d52cc5c9db7a81451826f15ec97eL239.

Maybe this can happen when a vm bound with a VIP